File: python/packages/ai/teams/ai/validators/json_response_validator.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from botbuilder.core import TurnContext
from jsonschema import ValidationError, validate

from ...state import MemoryBase
from ...utils.json import parse
from ..models.prompt_response import PromptResponse
from ..tokenizers import Tokenizer
from .prompt_response_validator import PromptResponseValidator
from .validation import Validation

NEW_LINE = "\n"


class JSONResponseValidator(PromptResponseValidator):
    """
    Default response validator that always returns true.
    """

    _schema: Optional[Dict[str, Any]]
    _missing_json_feedback: str
    _error_feedback: str

    def __init__(
        self,
        schema: Optional[Dict[str, Any]] = None,
        missing_json_feedback: Optional[str] = None,
        error_feedback: Optional[str] = None,
    ) -> None:
        """
        Creates a new `JSONResponseValidator` instance.
        """

        super().__init__()
        self._schema = schema
        self._missing_json_feedback = (
            missing_json_feedback
            if missing_json_feedback is not None
            else "No valid JSON objects were found in the response. Return a valid JSON object."
        )
        self._error_feedback = (
            error_feedback
            if error_feedback is not None
            else "The JSON returned had errors. Apply these fixes:"
        )

    @property
    def schema(self):
        """
        JSON schema to validate the response against.
        """
        return self._schema

    @property
    def missing_json_feedback(self):
        """
        Feedback given when no JSON is returned.
        """
        return self._missing_json_feedback

    @property
    def error_feedback(self):
        """
        Feedback prefix given when schema errors are detected.
        """
        return self._error_feedback

    async def validate_response(
        self,
        context: TurnContext,
        memory: MemoryBase,
        tokenizer: Tokenizer,
        response: PromptResponse[str],
        remaining_attempts: int,
    ) -> Validation:
        message = response.message

        if message is None:
            return Validation(valid=False)

        content = message.content if message.content is not None else ""
        parsed = parse(content)

        if len(parsed) == 0:
            return Validation(valid=False, feedback=self.missing_json_feedback)

        if self._schema is None:
            return Validation(value=parsed.pop())

        parsed.reverse()
        errors: List[str] = []

        for obj in parsed:
            try:
                validate(instance=obj, schema=self._schema)
                return Validation(value=obj)
            except ValidationError as err:
                errors.append(err.message)

        return Validation(
            valid=False, feedback=f"{self._error_feedback}{NEW_LINE}{NEW_LINE.join(errors)}"
        )

File: js/packages/teams-ai/src/validators/JSONResponseValidator.ts
import { TurnContext } from 'botbuilder';
import { Validator, Schema, ValidationError } from 'jsonschema';

import { Memory } from '../MemoryFork';
import { Validation, PromptResponseValidator } from './PromptResponseValidator';
import { Tokenizer } from '../tokenizers';
import { PromptResponse } from '../types';

import { Response } from './Response';

/**
 * Parses any JSON returned by the model and optionally verifies it against a JSON schema.
 * @template TValue Optional. Type of the validation value returned. Defaults to `Record<string, any>`.
 */
export class JSONResponseValidator<TValue = Record<string, any>> implements PromptResponseValidator<TValue> {
    /**
     * Creates a new `JSONResponseValidator` instance.
     * @param {Schema} schema Optional. JSON schema to validate the response against.
     * @param {string} missingJsonFeedback Optional. Custom feedback to give when no JSON is returned.
     * @param {string} errorFeedback Optional. Custom feedback prefix to use when schema errors are detected.
     */
    public constructor(schema?: Schema, missingJsonFeedback?: string, errorFeedback?: string) {
        this.schema = schema;
        this.missingJsonFeedback =
            missingJsonFeedback ?? 'No valid JSON objects were found in the response. Return a valid JSON object.';
        this.errorFeedback = errorFeedback ?? 'The JSON returned had errors. Apply these fixes:';
    }

    /**
     * Feedback prefix given when schema errors are detected.
     */
    public errorFeedback: string;

    /**
     * Feedback given when no JSON is returned.
     */
    public missingJsonFeedback: string;

    /**
     * JSON schema to validate the response against.
     */
    public readonly schema?: Schema;

    /**
     * Validates a response to a prompt.
     * @param {TurnContext} context Context for the current turn of conversation with the user.
     * @param {Memory} memory An interface for accessing state values.
     * @param {Tokenizer} tokenizer Tokenizer to use for encoding and decoding text.
     * @param {PromptResponse<string>} response Response to validate.
     * @param {number} remaining_attempts Number of remaining attempts to validate the response.
     * @returns {Promise<Validation<TValue>>} A `Validation` object.
     */
    public validateResponse(
        context: TurnContext,
        memory: Memory,
        tokenizer: Tokenizer,
        response: PromptResponse<string>,
        remaining_attempts: number
    ): Promise<Validation<TValue>> {
        const message = response.message!;
        const text = message.content ?? '';

        // Parse the response text
        const parsed = Response.parseAllObjects(text);
        if (parsed.length == 0) {
            return Promise.resolve({
                type: 'Validation',
                valid: false,
                feedback: this.missingJsonFeedback
            });
        }

        // Validate the response against the schema
        if (this.schema) {
            let errors: ValidationError[] | undefined;
            const validator = new Validator();
            for (let i = parsed.length - 1; i >= 0; i--) {
                const obj = Response.removeEmptyValuesFromObject(parsed[i]);
                const result = validator.validate(obj, this.schema);
                if (result.valid) {
                    return Promise.resolve({
                        type: 'Validation',
                        valid: true,
                        value: obj as TValue
                    });
                } else if (!errors) {
                    errors = result.errors;
                }
            }

            return Promise.resolve({
                type: 'Validation',
                valid: false,
                feedback: `${this.errorFeedback}\n${errors!.map((e) => this.getErrorFix(e)).join('\n')}`
            });
        } else {
            // Return the last object
            return Promise.resolve({
                type: 'Validation',
                valid: true,
                value: parsed[parsed.length - 1] as TValue
            });
        }
    }

    /**
     * @private
     * @param {ValidationError} error Error in the JSON object
     * @returns {string} How to fix the given error.
     */
    private getErrorFix(error: ValidationError): string {
        // Get argument as a string
        let arg: string;
        if (Array.isArray(error.argument)) {
            arg = error.argument.join(',');
        } else if (typeof error.argument === 'object') {
            arg = JSON.stringify(error.argument);
        } else {
            arg = error.argument.toString();
        }

        switch (error.name) {
            case 'type':
                // field is of the wrong type
                return `convert "${error.property}" to a ${arg}`;
            case 'anyOf':
                // field is not one of the allowed types
                return `convert "${error.property}" to one of the allowed types in the provided schema.`;
            case 'additionalProperties':
                // field has an extra property
                return `remove the "${arg}" property from ${
                    error.property == 'instance' ? 'the JSON object' : `"${error.property}"`
                }`;
            case 'required':
                // field is missing a required property
                return `add the "${arg}" property to ${
                    error.property == 'instance' ? 'the JSON object' : `"${error.property}"`
                }`;
            // TODO: jsonschema does not validate formats by default. https://github.com/microsoft/teams-ai/issues/1080
            case 'format':
                // field is not in the correct format
                return `change the "${error.property}" property to be a ${arg}`;
            case 'uniqueItems':
                // field has duplicate items
                return `remove all duplicate items from "${error.property}"`;
            case 'enum':
                // field is not one of the allowed values
                arg = error.message.split(':')[1].trim();
                return `change the "${error.property}" property to be one of these values: ${arg}`;
            case 'const':
                // field is not the correct value
                return `change the "${error.property}" property to be ${arg}`;
            default:
                return `"${error.property}" ${error.message}. Fix that`;
        }
    }
}
